recipes-extended/xen/files/shim-support-for-shim-lock-measure.patch

```diff
@@ -81,38 +77,68 @@ Signed-off-by: Tamas K Lengyel <lengyelt@ainfosec.com>
 +            }
 +            else
 +            {
++                /* Re-ordering measurements means seal-system needs tweaking? */
 +                if ( (status = shim_lock->Measure(kernel.ptr, kernel.size, 4))
 +                    != EFI_SUCCESS )
 +                    PrintErrMesg(L"Dom0 kernel image could not be measured",
 +                                 status);
 +            }
-+        }
++            kernel_already_measured = true;
+         }
  
-         name.s = get_value(&cfg, section.s, "ramdisk");
-         if ( name.s )
-         {
-             read_file(dir_handle, s2w(&name), &ramdisk, NULL);
-             efi_bs->FreePool(name.w);
+         if ( !read_section(loaded_image, L"ramdisk", &ramdisk, NULL) )
+@@ -1313,6 +1347,11 @@ efi_start(EFI_HANDLE ImageHandle, EFI_SY
+             {
+                 read_file(dir_handle, s2w(&name), &ramdisk, NULL);
+                 efi_bs->FreePool(name.w);
 +
-+            if ( shim_lock &&
-+                (status = shim_lock->Measure(ramdisk.ptr, ramdisk.size, 8))
-+                 != EFI_SUCCESS )
-+                    PrintErrMesg(L"ramdisk could not be measured", status);
++                if ( shim_lock &&
++                    (status = shim_lock->Measure(ramdisk.ptr, ramdisk.size, 8))
++                     != EFI_SUCCESS )
++                        PrintErrMesg(L"ramdisk could not be measured", status);
+             }
          }
  
-         name.s = get_value(&cfg, section.s, "xsm");
-@@ -1323,6 +1357,11 @@ efi_start(EFI_HANDLE ImageHandle, EFI_SY
-         {
-             read_file(dir_handle, s2w(&name), &xsm, NULL);
-             efi_bs->FreePool(name.w);
+@@ -1323,6 +1362,11 @@ efi_start(EFI_HANDLE ImageHandle, EFI_SY
+             {
+                 read_file(dir_handle, s2w(&name), &xsm, NULL);
+                 efi_bs->FreePool(name.w);
 +
-+            if ( shim_lock &&
-+                (status = shim_lock->Measure(xsm.ptr, xsm.size, 8))
-+                 != EFI_SUCCESS)
-+                    PrintErrMesg(L"XSM policy could not be measured", status);
++                if ( shim_lock &&
++                    (status = shim_lock->Measure(xsm.ptr, xsm.size, 8))
++                     != EFI_SUCCESS)
++                        PrintErrMesg(L"XSM policy could not be measured", status);
+             }
          }
  
-         /*
+@@ -1387,10 +1431,23 @@ efi_start(EFI_HANDLE ImageHandle, EFI_SY
+      * verify it.
+      */
+     if ( kernel.ptr &&
+-         !EFI_ERROR(efi_bs->LocateProtocol(&shim_lock_guid, NULL,
+-                                           (void **)&shim_lock)) &&
+-         (status = shim_lock->Verify(kernel.ptr, kernel.size)) != EFI_SUCCESS )
+-        PrintErrMesg(L"Dom0 kernel image could not be verified", status);
++         shim_lock && !kernel_already_measured) {
```

Comment:
Out of curiosity, why duplicate here if we know that our measurement has already happened? In what case would we be hitting this second block without a measured kernel?

Reply:
I think the re-ordering in https://github.com/OpenXT/xenclient-oe/pull/1461/commits/ad0555604b2f01b4ee3753869ea5b52d91332e95 is unnecessary.  I confused myself that the movement of the kernel measuring would change the PCR calculation.  However, moving the kernel is okay since it is measured into PCR 4 while the others are measured into PCR 8.  An HP laptop failed to unseal which made me think this was the problem, but that laptop has some other issue with PCR4.

I'm going to test out removing `kernel_already_measured` and just having the single kernel location.

Reply:
Forward seal & unseal failed when I removed kernel_already_measured.  I tested on a Dell is typically good for sealing & unsealing.

Reply:
I guess I'm still confused on the two cases then. The implication with that variable is that sometimes the kernel will already be measured at this point and sometimes it won't be. In what case will you have one vs the other? And why isn't it universal across all platforms?

Reply:
Well, I was trying to make it work for x86 or ARM.  Maybe I shouldn't bother with that and just stick to x86.

shim-support-for-shim-lock-measure.patch doesn't have a problem with the order of PCR4 & 8 extends, but tboot-measure-and-launch-from-xen-efi.patch extends tboot into PCR4.  So that is what throws off PCR4 xen.efi->tboot vs tboot->xen.efi

I can re-spin without kernel_already_measured and break the ARM devicetree path to just have the 1) xen.efi 2) tboot extend order.  Does that sound good?

Reply:
I would think ARM is a moot consideration seeing as how we don't support it for the larger usecase. Personally I would say to focus on x86 to avoid confusion. Curious to hear from @crogers1 though.

Reply:
xen commit 048de2c47075d6106136bee110d4e930049a0fe34.15.0-rc4-717-g048de2c470 , in 4.16, re-ordered the kernel measurement location in efi_start from inside the config file loop to after that loop and the device tree parsing.

With tboot-measure-and-launch-from-xen-efi.patch, a faux diff of pseudo code is roughly:
```
 efi_start
-    measure kernel
     measure things in PCR8
     arch_late
         measure tboot
         measure xen.efi
+    measure kernel
```


```
 pcr4_objs=(
     /boot/shimx64.efi
     /boot/xen.efi
-    /boot/bzImage
     /boot/tboot
     /boot/xen.efi
+    /boot/bzImage
 )
```

So I think the pcr4_objs should just be re-ordered and shim-support-for-shim-lock-measure.patch can just patch the new "measure kernel" location.  Since we forward seal from a chroot, we'll use the new pcr4 order and things should be fine.

This also makes pcr4 follow the actual execution order since the boot goes through xen -> tboot -> xen -> bzImage.

Reply:
+1 for aligning the order with upstream. Like you said forward seal should be unaffected.

Reply:
Also +1